adsmp/app.py

```diff
@@ -425,43 +378,84 @@ def metrics_session_scope(self):
         finally:
             s.close()        
 
+    def index_metrics(self, batch, priority=0, update_timestamps=True):
+        # todo: need to pass update_timestamps
+        success, metrics_exception = self.update_metrics_db(batch)
+        bibcodes = [x['bibcode'] for x in batch]
+        failed = set(bibcodes) - set(success)
+        self.index_complete(batch, failed, 'metrics')
+        return failed
 
-    def update_metrics_db(self, batch_insert, batch_update):
-        """Inserts data into the metrics DB.
-        :param: batch_insert - list of json objects to insert into the metrics
-        :param: batch_update - list of json objects to update in metrics db
+    def index_datalinks(self, links_data, priority=0, update_timestamps=True):
+        # todo is failed right?
+        links_url = self.conf.get('LINKS_RESOLVER_UPDATE_URL')
+        api_token = self.conf.get('ADS_API_TOKEN', '')
+        failed = []
+        if len(links_data):
+            bibcodes = [x['bibcode'] for x in links_data]
+            r = requests.put(links_url, data=json.dumps(links_data), headers={'Authorization': 'Bearer {}'.format(api_token)})
+            if r.status_code == 200:
+                self.logger.info('sent %s datalinks to %s including %s', len(links_data), links_url, links_data[0])
+            else:
+                self.logger.error('error sending links to %s, error = %s', links_url, r.text)
+                failed = bibcodes
+                self.mark_processed([x['bibcode'] for x in links_data], type=None, status='links-failed')
+        self.index_complete(links_data, failed, 'datalinks')
+        return failed
         
+    def index_complete(self, batch, failed, kind):
+        """update all state info in records database for last index
+
+        kind is one of: solr, metrics or datalinks
+        # consider bulk update https://stackoverflow.com/questions/25694234/bulk-update-in-sqlalchemy-core-using-where
+"""
+        if not self.update_timestamps:
+            return
+        bibcodes = [x['bibcode'] for x in batch]
+        bibcode_to_data = {doc['bibcode']: doc for doc in batch}
+        with self.session_scope() as session:
+            for bibcode in bibcodes:
+                c = self.checksum(bibcode_to_data[bibcode])
+                r = session.query(Records).filter_by(bibcode=bibcode) \
+                                          .options(_load_only('bibcode', kind + '_checksum',
+                                                              kind + '_processed', 'status')) \
+                                          .first()
+                setattr(r, kind + '_checksum', c)
+                if bibcode in failed:
+                    setattr(r, 'status', kind + '-failed')
+                elif getattr(r, 'status') == kind + '-failed':
+                    setattr(r, 'status', 'success')
+                setattr(r, kind + '_processed', adsputils.get_date())
+                session.commit()
```

Comment:
With the new parallel logic, it is going to be more common for the `status` field to be overwritten. For instance, if solr fails for a bibcode but then metrics and/or resolver succeeds, the end value of `status` is going to be success. Can we have three independent status fields instead? This would require also a change in our retrying mechanism, but it should not be big.

Reply:
I'll delete the status field and three new status fields, one for each production data store.
Regarding failure modes, I think your comment misunderstood the elif statement at line 426.  If solr fails for a bibcode and then metrics succeeds metrics won't overwrite and the status remains solr-failed.  Unfortunately, with only one status field the code doesn't handle other failure modes as well.  If solr fails and then metrics fails it overwrites solr-failed with metrics-failed.  This will be fixed with multiple status fields.

Reply:
Oh, I see! That's smart, I misinterpreted that logic indeed. And yeah, you are right, having multiple fields will allow us to see when multiple errors happen.